packages/core/src/browser/shell/tab-bar-toolbar.tsx

```diff
@@ -183,6 +207,7 @@ export class TabBarToolbar extends ReactWidget {
         if (TabBarToolbarItem.is(item)) {
             this.commands.executeCommand(item.command, this.current);
```

Comment:
One more minor point. The execution may return a promise, and if it does, the following `this.update()` may happen before the command has actually performed its work, which means it may be checking the state too soon.

 This is a slightly tricky case since a command can actually include user interaction and take quite a long time to complete, so it may just be best to pretend that it's synchronous and trust that if it has a delayed effect, the registerer of the toolbar item has had the foresight to include an `onDidChange` that will fire when they're done to trigger an update in the tabbar toolbar.

Reply:
After offline discussion, I think it's best not to worry about this, and leave the handling of any long-running promises to the registerer of the tabbar item. If they need to update the tabbar, they can include `onDidChange` and fire that event.